diff --git a/client/src/components/Markdown/Elements/HistoryDatasetAsTable.vue b/client/src/components/Markdown/Elements/HistoryDatasetAsTable.vue
new file mode 100644
index 000000000000..dd74936ca879
--- /dev/null
+++ b/client/src/components/Markdown/Elements/HistoryDatasetAsTable.vue
@@ -0,0 +1,126 @@
+<script setup lang="ts">
+import { computed } from "vue";
+
+import { UrlDataProvider } from "@/components/providers/UrlDataProvider.js";
+
+import LoadingSpan from "@/components/LoadingSpan.vue";
+
+interface HistoryDatasetAsTableProps {
+    historyDatasetId: string;
+    compact: boolean;
+    showColumnHeaders: boolean;
+    title?: string;
+    footer?: string;
+}
+
+const props = withDefaults(defineProps<HistoryDatasetAsTableProps>(), {
+    compact: false,
+    showColumnHeaders: true,
+    title: undefined,
+    footer: undefined,
+});
+
+const itemUrl = computed(() => {
+    return `/api/datasets/${props.historyDatasetId}/get_content_as_text`;
+});
+
+const metaUrl = computed(() => {
+    return `/api/datasets/${props.historyDatasetId}`;
+});
+
+const expanded = false;
+
+const contentClass = computed(() => {
+    if (expanded) {
+        return "embedded-dataset-expanded";
+    } else {
+        return "embedded-dataset";
+    }
+});
+
+function getFields(metaData: any) {
+    const fields = [];
+    const columnNames = metaData.metadata_column_names || [];
+    const columnCount = metaData.metadata_columns;
+    for (let i = 0; i < columnCount; i++) {
+        fields.push({
+            key: `${i}`,
+            label: columnNames[i] || i,
+            sortable: true,
+        });
+    }
+    return fields;
+}
+
+function getItems(textData: any, metaData: any) {
+    const tableData: object[] = [];
+    const delimiter: string = metaData.metadata_delimiter || "\t";
+    const comments: number = metaData.metadata_comment_lines || 0;
+    const lines = textData.split("\n");
+    lines.forEach((line: string, i: number) => {
+        if (i >= comments) {
+            const tabs = line.split(delimiter);
+            const rowData: string[] = [];
+            let hasData = false;
+            tabs.forEach((cellData: string, j: number) => {
+                const cellDataTrimmed = cellData.trim();
+                if (cellDataTrimmed) {
+                    hasData = true;
+                }
+                rowData[j] = cellDataTrimmed;
+            });
+            if (hasData) {
+                tableData.push(rowData);
+            }
+        }
+    });
+    return tableData;
+}
+</script>
+
+<template>
+    <b-card :no-body="props.compact">
+        <b-card-title v-if="title">
+            <b>{{ title }}</b>
+        </b-card-title>
+        <UrlDataProvider v-slot="{ result: itemContent, loading, error }" :url="itemUrl">
+            <LoadingSpan v-if="loading" message="Loading Dataset" />
+            <div v-else-if="error">{{ error }}</div>
+            <div v-else :class="contentClass">
+                <div v-if="itemContent.item_data">
+                    <div>
+                        <UrlDataProvider
+                            v-slot="{ result: metaData, loading: metaLoading, error: metaError }"
+                            :url="metaUrl">
+                            <LoadingSpan v-if="metaLoading" message="Loading Metadata" />
+                            <div v-else-if="metaError">{{ metaError }}</div>
+                            <b-table
+                                v-else
+                                :thead-class="props.showColumnHeaders ? '' : 'd-none'"
+                                striped
+                                hover
+                                :fields="getFields(metaData)"
+                                :items="getItems(itemContent.item_data, metaData)" />
+                        </UrlDataProvider>
+                    </div>
+                </div>
+                <div v-else>No content found.</div>
+                <b-link v-if="itemContent.truncated" :href="itemContent.item_url"> Show More... </b-link>
+            </div>
+        </UrlDataProvider>
+        <b-card-footer v-if="footer">
+            {{ footer }}
+        </b-card-footer>
+    </b-card>
+</template>
+
+<style scoped>
+.embedded-dataset {
+    max-height: 20rem;
+    overflow-y: auto;
+}
+.embedded-dataset-expanded {
+    max-height: 40rem;
+    overflow-y: auto;
+}
+</style>
diff --git a/client/src/components/Markdown/MarkdownContainer.vue b/client/src/components/Markdown/MarkdownContainer.vue
index c338743a39e2..13806a855da4 100644
--- a/client/src/components/Markdown/MarkdownContainer.vue
+++ b/client/src/components/Markdown/MarkdownContainer.vue
@@ -4,6 +4,7 @@ import { computed, ref } from "vue";
 import { useConfig } from "@/composables/config";
 
 import HistoryDatasetAsImage from "./Elements/HistoryDatasetAsImage.vue";
+import HistoryDatasetAsTable from "./Elements/HistoryDatasetAsTable.vue";
 import HistoryDatasetCollectionDisplay from "./Elements/HistoryDatasetCollection/CollectionDisplay.vue";
 import HistoryDatasetDetails from "./Elements/HistoryDatasetDetails.vue";
 import HistoryDatasetDisplay from "./Elements/HistoryDatasetDisplay.vue";
@@ -68,6 +69,17 @@ const props = defineProps({
 
 const isCollapsible = computed(() => props.args.collapse !== undefined);
 const isVisible = computed(() => !isCollapsible.value || toggle.value);
+
+function argToBoolean(args, name, booleanDefault) {
+    const valueAsString = args[name];
+    if (valueAsString == "true") {
+        return true;
+    } else if (valueAsString == "false") {
+        return false;
+    } else {
+        return booleanDefault;
+    }
+}
 </script>
 
 <template>
@@ -126,6 +138,13 @@ const isVisible = computed(() => !isCollapsible.value || toggle.value);
             </InstanceUrl>
             <HistoryLink v-else-if="name == 'history_link'" :args="args" :histories="histories" />
             <HistoryDatasetAsImage v-else-if="name == 'history_dataset_as_image'" :args="args" />
+            <HistoryDatasetAsTable
+                v-else-if="name == 'history_dataset_as_table'"
+                :history-dataset-id="args.history_dataset_id"
+                :compact="argToBoolean(args, 'compact', false)"
+                :show-column-headers="argToBoolean(args, 'show_column_headers', true)"
+                :title="args.title"
+                :footer="args.footer" />
             <HistoryDatasetLink v-else-if="name == 'history_dataset_link'" :args="args" />
             <HistoryDatasetIndex v-else-if="name == 'history_dataset_index'" :args="args" />
             <InvocationTime v-else-if="name == 'invocation_time'" :args="args" :invocations="invocations" />
diff --git a/client/src/components/Markdown/MarkdownToolBox.vue b/client/src/components/Markdown/MarkdownToolBox.vue
index cad2605a50f5..e7f2d7c6b1ca 100644
--- a/client/src/components/Markdown/MarkdownToolBox.vue
+++ b/client/src/components/Markdown/MarkdownToolBox.vue
@@ -77,6 +77,11 @@ const historySharedElements = [
         name: "Embedded Dataset",
         emitter: "onHistoryDatasetId",
     },
+    {
+        id: "history_dataset_as_table",
+        name: "Embedded Dataset (as table)",
+        emitter: "onHistoryDatasetId",
+    },
     {
         id: "history_dataset_type",
         name: "Dataset Type",
diff --git a/lib/galaxy/managers/markdown_parse.py b/lib/galaxy/managers/markdown_parse.py
index ae70de230f91..4d0f20b5d2ca 100644
--- a/lib/galaxy/managers/markdown_parse.py
+++ b/lib/galaxy/managers/markdown_parse.py
@@ -30,6 +30,16 @@ class DynamicArguments:
     "history_dataset_display": ["input", "output", "history_dataset_id"],
     "history_dataset_embedded": ["input", "output", "history_dataset_id"],
     "history_dataset_as_image": ["input", "output", "history_dataset_id", "path"],
+    "history_dataset_as_table": [
+        "input",
+        "output",
+        "history_dataset_id",
+        "path",
+        "title",
+        "footer",
+        "show_column_headers",
+        "compact",
+    ],
     "history_dataset_peek": ["input", "output", "history_dataset_id"],
     "history_dataset_info": ["input", "output", "history_dataset_id"],
     "history_dataset_link": ["input", "output", "history_dataset_id", "path", "label"],
diff --git a/lib/galaxy/managers/markdown_util.py b/lib/galaxy/managers/markdown_util.py
index 3523031283cf..9ef57141375e 100644
--- a/lib/galaxy/managers/markdown_util.py
+++ b/lib/galaxy/managers/markdown_util.py
@@ -137,6 +137,10 @@ def _remap(container, line):
                 _check_object(object_id, line)
                 hda = hda_manager.get_accessible(object_id, trans.user)
                 rval = self.handle_dataset_as_image(line, hda)
+            elif container == "history_dataset_as_table":
+                _check_object(object_id, line)
+                hda = hda_manager.get_accessible(object_id, trans.user)
+                rval = self.handle_dataset_as_table(line, hda)
             elif container == "history_dataset_peek":
                 _check_object(object_id, line)
                 hda = hda_manager.get_accessible(object_id, trans.user)
@@ -251,6 +255,10 @@ def handle_dataset_display(self, line, hda):
     def handle_dataset_as_image(self, line, hda):
         pass
 
+    @abc.abstractmethod
+    def handle_dataset_as_table(self, line, hda):
+        pass
+
     @abc.abstractmethod
     def handle_dataset_peek(self, line, hda):
         pass
@@ -411,6 +419,9 @@ def handle_history_link(self, line, history):
     def handle_dataset_as_image(self, line, hda):
         pass
 
+    def handle_dataset_as_table(self, line, hda):
+        pass
+
     def handle_job_metrics(self, line, job):
         pass
 
@@ -538,6 +549,14 @@ def _embed_image(self, name: str, image_type: str, image_data: bytes):
         base64_image_data = base64.b64encode(image_data).decode("utf-8")
         return f"![{name}](data:image/{image_type};base64,{base64_image_data})"
 
+    def handle_dataset_as_table(self, line, hda):
+        # TODO: this form of the rendering doesn't do anything special with advanced
+        # options yet but could easily be modified in the future. show_column_headers,
+        # compact, title, and footer should be handled in here to bring the PDF and the
+        # web rendering closer.
+        rval = self.handle_dataset_embedded(line, hda)
+        return rval
+
     def handle_history_link(self, line, history):
         if history:
             content = literal_via_fence(history.name)
